Выполнить ФЗ22. Реализовать Сервисы Order, Payment, Delivery. Организ…

Volckoff opened the pull request:
Реализовать Сервисы Order, Payment, Delivery. Организовать Межсервисное взаимодействие через Feign. Организовать внешнее общение сервисов через Gateway.

Volckoff committed:
Выполнить ФЗ22. Реализовать Сервисы Order, Payment, Delivery. Организовать Межсервисное взаимодействие через Feign. Организовать внешнее общение сервисов через Gateway.

Volckoff commented:
Александр, добрый день!

По итогу проект вышел действительно интересным )
В этот раз от mapStruct я сразу отказался (мапперы совсем небольшие).
Выучили много полезных фич, которые нам с предыдущими проектами жизнь упростили бы.
Постарался учесть все предыдущие ошибки которые совершал, чтобы не повторяться.
Было добавлено несколько эндпоинтов о которых в тз не говорилось, но наставник сказал что это нормально. И вообще как выяснилось эта спецификация и к предыдущему заданию у нас не совсем корректные, например в одном из эндпоинтов у нас прилетает запрос параметрами а в спецификации указано RequestBody.
Gateway мне понравился - крайне удобная вещь, конфиг сервер тоже в принципе ... однако как решить проблему точки уязвимости кроме запуска нескольких экземпляров сервиса я пока не понял (я имею ввиду что если например кофиг сервер отваливается, то ни один микросервис не запустится). Архитектура приложения конечно далека от идеала, но это с опытом как мне кажется.
Было принято решение с Cash и SpringSecurity поэкспериментировать позже, чтобы не затягивать сроки, да и в тз они не указаны (про Cash нам вообще только на вебинаре рассказали, хотя я так понимаю в продакшн коде, при использовании REST, он почти всегда используется). 
Так же в микросервисном приложении запуск довольно капризная вещь, т.к. некоторые сервисы приходится ждать (например пока сервис зарегистрируется в Eureka или подтянет Config), а как запустить все разом с нужными таймингами я не разобрался, наставник сказал что вообще заморачиваться не стоит (но каждый раз запускать хотя бы 9 микросервисов, чтобы протестировать функциональность - неудобно, с учетом того что иногда приходится ронять контейнер, чтобы база очистилась).

avfyodorov commented on commerce/delivery/src/main/java/ru/practicum/service/DeliveryServiceImpl.java:
```diff
@@ -0,0 +1,143 @@
+        Delivery delivery = getDeliveryByOrderId(orderDto.getOrderId());
+        log.info("Delivery for calc cost: {}", delivery);
+
+        double cost = DeliveryUtil.BASE_DELIVERY_PRICE;
```
Всё-таки, деньги лучше считать в BigDecimal, а не double

Volckoff replied:
Исправлено:
Денежные расчеты производятся в BigDecimal.

avfyodorov commented on commerce/delivery/src/main/java/ru/practicum/service/DeliveryServiceImpl.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    @Override
+    public Double deliveryCost(OrderDto orderDto) {
```
Вот тут выполняется какой-то  расчет. С коэффициентами. 0-) У нас работает микросервисное приложение, туча запросов и где-то что-то неправильно посчиталось, разбираться будет очень трудно.
Есть предложение - залогировать основные моменты, что пришло, что получилось. 
И везде указать по какому заказу, его id, чтобы записи можно было идентифицировать.

Volckoff replied:
Исправлено:
Добавлено подробное логирование с ID заказа:
Все логи содержат orderId для идентификации записей в приложении.

avfyodorov reviewed (commented):
Добрый день, Владимир! 

В целом всё в порядке, единственное - расчёт стоимости может быть имеет смысл доработать

Volckoff committed:
Внести правки согласно комментариям ревью:

- Денежные расчеты производить в BigDecimal:
Обновить DeliveryUtil — все константы теперь BigDecimal
Обновить DeliveryServiceImpl.deliveryCost() — все расчеты используют BigDecimal
Обновить интерфейсы: DeliveryService, DeliveryOperation
Обновить контроллеры: DeliveryController
Обновить сервисы: OrderFacadeImpl, OrderServiceImpl для работы с BigDecimal

- Добавить подробное логирование с ID заказа:
Все логи содержат orderId для идентификации записей в приложении.

Volckoff commented:
Александр, спасибо за грамотное замечание:
Немного изучив тему, я узнал что использование BigDecimal - bestPractice для финансовых расчетов, т.к. не имеет погрешности в отличие от double. Так же попробовал применить RoundingMode для установки округления цифр после запятой.
Очень занятная библиотека.

avfyodorov reviewed (commented):
Добрый вечер, Владимир!

Замечаний нет.
Работа принята.
